File: node/src/EnvConfigLoaderFactory.ts
import GithubEnvConfigLoader from './GithubEnvConfigLoader';
import IEnvConfigLoader from './IEnvConfigLoader';

/**
 * Config loader factory.
 * Should one decide working with config that is different than github, a proper concrete impl should be provided,
 * the getLoader method should return an instance to the alternative loader (maybe gitlab or something else)
 * In case this library will provide multiple options, the instantiation should be made dynamically from config (env, arg)
 */
export default class EnvConfigLoaderFactory {
    public getLoader(name = 'default'): IEnvConfigLoader {
        const loadersMap: any = {
            github: GithubEnvConfigLoader,
            default: GithubEnvConfigLoader,
        };

        const loader = loadersMap[name];

        return new loader(); // eslint-disable-line
    }
}

File: ruby/src/env_config_loader_factory.rb
require_relative './github_env_conf_loader.rb'

LOADERS_MAP = {
  'github' => GithubEnvConfigLoader,
  'default' => GithubEnvConfigLoader
}

class EnvConfigLoaderFactory
  # """
  # The most naive and yet decoupled factory for config loader.
  # In case someone changes the implementation to gitlab or something else,
  # return the new implementation instance without affecting the rest of the code
  #
  # @return [EnvConfLoader] concrete instance -- the sought after loader
  def get_loader(loader_name = nil)
    loader = LOADERS_MAP['default']

    unless loader_name.nil?
      loader = LOADERS_MAP[loader_name]
    end

    loader.new
  end
end
